neps/nep-0393.md

```diff
@@ -0,0 +1,480 @@
+    pub expires_at: Option<u64>, // When token expires, Unix epoch in milliseconds
+    pub reference: Option<String>, // URL to an off-chain JSON file with more info.
+    pub reference_hash: Option<Base64VecU8>, // Base64-encoded sha256 hash of JSON from reference field. Required if `reference` is included.
+}
+
+
+trait SBTRegistry {
+    /**********
+    * QUERIES
+    **********/
+
+    /// Get the information about specific token ID issued by `ctr` SBT contract.
+    fn sbt(&self, ctr: AccountId, token: TokenId) -> Option<Token>;
+
+    /// Returns total amount of tokens issued by `ctr` SBT contract, including expired tokens.
+    /// Depending on the implementation, if a revoke removes a token, it then is should not be
+    /// included in the supply.
+    fn sbt_supply(&self, ctr: AccountId) -> u64;
+
+    /// Returns total amount of tokens of given class minted by `ctr`. See `sbt_supply` for
+    /// information about revoked tokens.
+    fn sbt_supply_by_class(&self, ctr: AccountId, class: ClassId) -> u64;
+
+    /// Returns total supply of SBTs for a given owner. See `sbt_supply` for information about
+    /// revoked tokens.
+    /// If class is specified, returns only owner supply of the given class -- must be 0 or 1.
+    fn sbt_supply_by_owner(
+        &self,
+        ctr: AccountId,
+        account: AccountId,
+        class: Option<ClassId>,
+    ) -> u64;
+
+    /// Query sbt tokens issued by a given contract.
+    /// If `from_index` is not specified, then `from_index` should be assumed
+    /// to be the first valid token id.
+    fn sbt_tokens(
+        &self,
+        ctr: AccountId,
+        from_index: Option<u64>,
+        limit: Option<u32>,
+    ) -> Vec<TokenId>;
+
+    /// Query SBT tokens by owner
+    /// If `from_class` is not specified, then `from_class` should be assumed to be the first
+    /// valid class id.
+    /// Returns list of pairs: `(Contract address, list of token IDs)`.
+    fn sbt_tokens_by_owner(
+        &self,
+        account: AccountId,
+        ctr: Option<AccountId>,
+        from_class: Option<u64>,
+        limit: Option<u32>,
+    ) -> Vec<(AccountId, Vec<TokenId>)>;
+
+    /// checks if an `account` was banned by the registry.
+    fn is_banned(&self, account: AccountId) -> bool;
+
+
+    /*************
+     * Transactions
+     *************/
+
+    /// Creates a new, unique token and assigns it to the `receiver`.
+    /// `token_spec` is a vector of pairs: owner AccountId and TokenMetadata.
+    /// Each TokenMetadata must have non zero `class`.
+    /// Must be called by an SBT contract.
+    /// Must emit `Mint` event.
+    /// Must provide enough NEAR to cover registry storage cost.
+    // #[payable]
+    fn sbt_mint(&mut self, token_spec: Vec<(AccountId, TokenMetadata)>) -> Vec<TokenId>;
+
+    /// sbt_recover reassigns all tokens from the old owner to a new owner,
+    /// and registers `old_owner` to a burned addresses registry.
```

Comment:
Using the word `burn` for both accounts and tokens is confusing.

According to this spec no `Ban` events will be emitted here. But the only `Burn` event defined in this spec deals with tokens. Can we please rephrase?

Also see my concern regarding a potential privilege escalation attack vector, should this call `Ban` `old_owner` registry-wide.